tests/transforms/test_specs.py

```diff
@@ -0,0 +1,165 @@
+
+        info = info_func(x, y, add_RY=False)
+
+        circuit(x, y, add_RY=False)
+
+        assert info == circuit.specs
+
+        assert len(info) == len_info
+
+        assert info["gate_sizes"] == defaultdict(int, {1: 2, 3: 1, 2: 1})
+        assert info["gate_types"] == defaultdict(int, {"RX": 1, "Toffoli": 1, "CRY": 1, "Rot": 1})
+        assert info["total_operations"] == 4
+        assert info["total_observables"] == 2
+        assert info["total_diagonalizing_gates"] == 1
+        assert info["num_used_wires"] == 3
+        assert info["depth"] == 3
+        assert info["num_device_wires"] == 4
+        assert info["diff_method"] == diff_method
+
+        if diff_method == "parameter-shift":
+            assert info["num_parameter_shift_executions"] == 7
+
+        if diff_method != "backprop":
+            assert info["device_name"] == "default.qubit"
+            assert info["num_trainable_params"] == 4
+        else:
+            assert info["device_name"] == "default.qubit.autograd"
+
+    @pytest.mark.parametrize(
+        "diff_method, len_info", [("backprop", 10), ("parameter-shift", 11), ("adjoint", 11)]
+    )
+    def test_specs_state(self, diff_method, len_info):
+        """Test specs works when state returned"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, diff_method=diff_method)
+        def circuit():
+            return qml.state()
+
+        info_func = qml.specs(circuit)
+        info = info_func()
+
+        circuit()
+        assert info == circuit.specs
+        assert len(info) == len_info
+
+        assert info["total_observables"] == 1
+        assert info["total_diagonalizing_gates"] == 0
+
+    def test_max_expansion(self):
+        """Test that a user can calculation specifications for a different max
+        expansion parameter."""
+
+        n_layers = 2
+        n_wires = 5
+
+        dev = qml.device("default.qubit", wires=n_wires)
+
+        @qml.qnode(dev)
+        def circuit(params):
+            qml.templates.BasicEntanglerLayers(params, wires=range(n_wires))
+            return qml.expval(qml.PauliZ(0))
+
+        params_shape = qml.templates.BasicEntanglerLayers.shape(n_layers=n_layers, n_wires=n_wires)
+        rng = np.random.default_rng(seed=10)
+        params = rng.standard_normal(params_shape)
+
+        assert circuit.max_expansion == 10
+        info = qml.specs(circuit, max_expansion=0)(params)
+        assert circuit.max_expansion == 10
+
+        assert len(info) == 10
+
+        assert info["gate_sizes"] == defaultdict(int, {5: 1})
+        assert info["gate_types"] == defaultdict(int, {"BasicEntanglerLayers": 1})
+        assert info["total_operations"] == 1
+        assert info["total_observables"] == 1
+        assert info["num_used_wires"] == 5
+        assert info["depth"] == 1
+        assert info["num_device_wires"] == 5
+        assert info["device_name"] == "default.qubit.autograd"
+        assert info["diff_method"] == "backprop"
```

Comment:
My only worry would be if there's any more edge cases, but the tests pass and coverage is 100% so that should be good!